packages/swap/src/quoting/Quoter.ts

```diff
@@ -0,0 +1,234 @@
+    });
+  }
+
+  private async quoteLegs(legs: [Leg]): Promise<[bigint]>;
+  private async quoteLegs(legs: [Leg, Leg]): Promise<[bigint, bigint]>;
+  private async quoteLegs(legs: [Leg, Leg | null]): Promise<[bigint, bigint | null]>;
+  private async quoteLegs([leg1, leg2]: [Leg] | [Leg, Leg | null]): Promise<
+    [bigint] | [bigint, bigint | null]
+  > {
+    const requestId = this.createId();
+
+    const requestLegs = [leg1.toMarketMakerJSON()] as
+      | [MarketMakerLeg]
+      | [MarketMakerLeg, MarketMakerLeg];
+
+    if (leg2) requestLegs[1] = leg2.toMarketMakerJSON();
+
+    const quoteRequest: MarketMakerQuoteRequest = { request_id: requestId, legs: requestLegs };
+
+    this.io.emit('quote_request', quoteRequest);
+
+    const quotes = await this.collectMakerQuotes(requestId);
+
+    const legLimitOrders = [leg1.toSwapInput()] as [SwapInput] | [SwapInput, SwapInput];
+
+    if (leg2) legLimitOrders[1] = leg2.toSwapInput();
+
+    for (const { legs } of quotes) {
+      for (let i = 0; i < legs.length; i += 1) {
+        legLimitOrders[i].limitOrders.push(...legs[i].map(([tick, amount]) => ({ tick, amount })));
+      }
+    }
+
+    const results = [getPriceFromQuotesAndPool(leg1, legLimitOrders[0])] as
+      | [Promise<bigint>]
+      | [Promise<bigint>, Promise<bigint> | null];
+
+    results[1] =
+      leg2 && legLimitOrders[1] ? getPriceFromQuotesAndPool(leg2, legLimitOrders[1]) : null;
+
+    return Promise.all(results);
+  }
+
+  async getQuote(
+    srcAsset: InternalAsset,
+    destAsset: InternalAsset,
+    swapInputAmount: bigint,
+    pools: Pool[],
+  ) {
+    const networkFee = getPoolsNetworkFeeHundredthPips(env.CHAINFLIP_NETWORK);
+
+    let intermediateAmount: bigint | null = null;
+    let outputAmount: bigint;
+    let networkFeeUsdc: bigint | null = null;
+
+    const fees: SwapFee[] = [];
+
+    if (srcAsset === 'Usdc' || destAsset === 'Usdc') {
+      if (srcAsset === 'Usdc') {
+        networkFeeUsdc = getHundredthPipAmountFromAmount(swapInputAmount, networkFee);
+      }
+
+      const leg = Leg.of(srcAsset, destAsset, swapInputAmount - (networkFeeUsdc ?? 0n));
+      [outputAmount] = await this.quoteLegs([leg]);
+
+      if (destAsset === 'Usdc') {
+        networkFeeUsdc = getHundredthPipAmountFromAmount(outputAmount, networkFee);
+        outputAmount -= networkFeeUsdc!;
+      }
+    } else {
+      let approximateUsdcAmount = await approximateIntermediateOutput(
+        srcAsset,
+        swapInputAmount.toString(),
+      );
+
+      networkFeeUsdc =
+        approximateUsdcAmount && getHundredthPipAmountFromAmount(approximateUsdcAmount, networkFee);
+
+      if (networkFeeUsdc && approximateUsdcAmount) {
+        approximateUsdcAmount -= networkFeeUsdc;
+      }
+
+      const leg1 = Leg.of(srcAsset, 'Usdc', swapInputAmount);
+      let leg2 = Leg.of('Usdc', destAsset, approximateUsdcAmount);
+
+      const quotes = await this.quoteLegs([leg1, leg2]);
+
+      if (
```

Comment:
im afraid a 1% difference is quite a lot to ignore 😕 we could interpolate the quote we got for the second leg to the output of the first leg

Reply:
not sure how to explain best what i mean (and interpolate is probable the wrong word):
1. we request a quote for `1 btc to usdc` and `60_000 usdc to eth` (bc we estimated an intermediate amount of 60000)
2. we recieve quotes: `55_000 usdc for 1 btc` and `20 eth for 60_000 usdc`
3. we assume the rate of the second quote for the output of the first: `20 / 60_000 * 55_000 = 18.33 ETH`

Reply:
i think with this we could even increase the threshold to something like 5%

Reply:
it's meant to be configurable, will add the env var. We'll see how fast the double round trip is in practice and optimize from there

Reply:
okay but can we decrease the default threshold at least? 
considering an btc->eth swap, this might give the user a better output amount than what was quoted by the market makers (we assume a 1% better usdc amount). atm for a 1 btc swap, the quote we give to the user can be 600$ better than what is realistic